File: Day-4/pub-sub-sample/python/event_bus.py
from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from threading import Lock
from typing import Callable


Handler = Callable[[str, dict], None]


@dataclass
class EventBus:
    _subscribers: dict[str, list[Handler]] = field(default_factory=lambda: defaultdict(list), init=False)
    _lock: Lock = field(default_factory=Lock, init=False)

    def subscribe(self, topic: str, handler: Handler) -> None:
        if not topic:
            raise ValueError("Topic must not be empty")

        with self._lock:
            self._subscribers[topic].append(handler)

    def publish(self, topic: str, payload: dict) -> int:
        with self._lock:
            handlers = list(self._subscribers.get(topic, []))

        for handler in handlers:
            handler(topic, payload)

        return len(handlers)

    def topic_snapshot(self) -> dict[str, int]:
        with self._lock:
            return {topic: len(handlers) for topic, handlers in self._subscribers.items()}

File: Day-4/pub-sub-sample/java/EventBus.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class EventBus {
    @FunctionalInterface
    public interface EventHandler {
        void handle(String topic, Map<String, Object> payload);
    }

    private final Map<String, List<EventHandler>> subscribers;

    public EventBus() {
        this.subscribers = new HashMap<>();
    }

    public synchronized void subscribe(String topic, EventHandler handler) {
        if (topic == null || topic.isBlank()) {
            throw new IllegalArgumentException("Topic must not be empty");
        }

        subscribers.computeIfAbsent(topic, key -> new ArrayList<>()).add(handler);
    }

    public int publish(String topic, Map<String, Object> payload) {
        List<EventHandler> handlers;
        synchronized (this) {
            handlers = new ArrayList<>(subscribers.getOrDefault(topic, List.of()));
        }

        for (EventHandler handler : handlers) {
            handler.handle(topic, payload);
        }

        return handlers.size();
    }

    public synchronized Map<String, Integer> topicSnapshot() {
        Map<String, Integer> snapshot = new LinkedHashMap<>();
        for (Map.Entry<String, List<EventHandler>> entry : subscribers.entrySet()) {
            snapshot.put(entry.getKey(), entry.getValue().size());
        }
        return snapshot;
    }
}
